sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -66,6 +66,8 @@ case class AssertTrue(child: Expression) extends UnaryExpression with ImplicitCa
 
   override def nullable: Boolean = true
 
+  override lazy val deterministic: Boolean = false
```

Comment:
hmm, this two expressions are very commonly used in other expressions' children. Making them non-deterministic will make such expressions used them as non-deterministic too, e.g., encoders. It might have big impact.

Reply:
yes, that's true. The change may be big indeed. I am open to other proposals, as creating a new `sideEffect` flag in the `Expression`.

Reply:
Because of this, I'm leaning towards creating a new flag instead of making them non-deterministic.

Reply:
I agree, thanks you all the comments. I'll try and add a new flag. I'll update the PR as soon as I have a solution for that. Probably it will take some time, though, as I will have to recheck all the usage of `determistic`.... Thanks.

Reply:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFAssertTrue.java#L45 Hive also marks it as non-deterministic. 

cc @maryannxue  too

Reply:
any comment on this @gatorsmile ?